### ISSUE TITLE
Pending Transaction Performance Problems

### ISSUE THREAD

**Issue by whymarrh** [MetaMask/metamask-extension#8377]
This issue details a possible performance issue that exists with pending transactions are handled.

Setup:

1. Load MetaMask
1. Onboarding, create wallet, get Ropsten Ether, etc.
1. Enable "Customize transaction nonce"
    - Under _Advanced Settings_
1. Copy account address
1. Send transaction to the same account **with nonce of 1** (pictured below)
    - This is supposed to be incorrect
1. Confirm that transaction
1. The transaction should be pending

<details>
<summary>Screenshot</summary>
<img width="1121" alt="" src="https://user-images.githubusercontent.com/1623628/79929731-22272780-8421-11ea-8424-c8487eeb4a6e.png">
</details>

### On `master` @ [`a7bcc17f9`](https://github.com/MetaMask/metamask-extension/tree/a7bcc17f9)/`v7.7.8`

Result:

- The transaction is stuck in pending (as expected)
- The `txMeta.history` for the transaction will grow unbounded at two rates:
    - Regularly/**slowly** as the transaction is retried with exponential backoff
        - 1 new entry each _((2^x) * (block time in seconds))_ seconds
        - An example `txMeta.history`: [`v778-slow-history.log`](https://github.com/MetaMask/metamask-extension/files/4513268/v778-slow-history.log)

    - **Rapidly** if the call to `eth_getTransactionReceipt` fails/throws
        - e.g. blocking the URL in devtools (pictured below)
            - [`https://api.infura.io/v1/jsonrpc/ropsten/eth_getTransactionReceipt?params=%5B%220x…%22%5D`](https://api.infura.io/v1/jsonrpc/ropsten/eth_getTransactionReceipt?params=%5B%220xcaa94e9fe1343bd5c1ec4b6440c4f64d0ca30527641a7562a7d442f2b004a78a%22%5D)
        - An example `txMeta.history`: [`v778-fast-history.log`](https://github.com/MetaMask/metamask-extension/files/4513339/v778-fast-history.log)

<details>
<summary>Screenshots</summary>
<img width="1561" alt="" src="https://user-images.githubusercontent.com/1623628/79929993-ee003680-8421-11ea-8054-0fda34dbb986.png">
<img width="1561" alt="" src="https://user-images.githubusercontent.com/1623628/79929995-ee98cd00-8421-11ea-8103-3f137e442027.png">
</details>

Once this history has grown significantly large, snapshotting the `txMeta` becomes prohibitvely expensive. (This is v7.7.8, before we added #8363.)

The rapid case—the worse one—is primarily the following loop:

- Update a transaction via `updateTx(…)`
- The transaction update is committed to memory via `_saveTxList` (`tx-state-manager.js:466`)
- `TransactionController#_updateMemstore` is run via `this.txStateManager.store.subscribe(…)` (`index.js:111`)
- Pending txs are updated via `updatePendingTxs` (`pending-tx-tracker.js:43`)
- A pending tx update fails, emitting a warning via `this.emit('tx:warning', txMeta, err)` (`pending-tx-tracker.js:186`)
- `updateTx(…)` is called via `this.pendingTxTracker.on('tx:warning', …)` (`index.js:613`)
- Rinse, repeat

<details>
<summary>Full stack trace</summary>
<p>
<a href="https://github.com/MetaMask/metamask-extension/files/4513292/network-fail-778-stacktrace-pathological.txt"><code>v778-network-fail-stacktrace-pathological.txt</code></a>

```
updateTx (tx-state-manager.js:195) // Line 195 here is simply where I had the breakpoint
(anonymous) (index.js:613) // this.pendingTxTracker.on('tx:warning', ...)
safeApply (index.js:70)
SafeEventEmitter.emit (index.js:56)
_checkPendingTx (pending-tx-tracker.js:186) // this.emit('tx:warning', txMeta, err)

--- await in _checkPendingTx (async) ---

(anonymous) (pending-tx-tracker.js:43) // await Promise.all(pendingTxs.map((txMeta) => this._checkPendingTx(txMeta)))
updatePendingTxs (pending-tx-tracker.js:43)

--- await in updatePendingTxs (async) ---

_updateMemstore (index.js:727)
(anonymous) (index.js:111) // this.txStateManager.store.subscribe(() => this._updateMemstore())
safeApply (index.js:70)
SafeEventEmitter.emit (index.js:62)
putState (index.js:22)
updateState (index.js:30)
_saveTxList (tx-state-manager.js:466)
updateTx (tx-state-manager.js:205)
(anonymous) (index.js:613) // this.pendingTxTracker.on('tx:warning', ...)
safeApply (index.js:70)
SafeEventEmitter.emit (index.js:56)
_checkPendingTx (pending-tx-tracker.js:186)

--- await in _checkPendingTx (async) ---

(anonymous) (pending-tx-tracker.js:43) // await Promise.all(pendingTxs.map((txMeta) => this._checkPendingTx(txMeta)))
updatePendingTxs (pending-tx-tracker.js:43)

--- await in updatePendingTxs (async) ---

_onLatestBlock (index.js:712)
safeApply (index.js:70)
SafeEventEmitter.emit (index.js:62)
_setCurrentBlock (base.js:131)
_newPotentialLatest (base.js:125)
_updateLatestBlock (polling.js:65)

--- await in _updateLatestBlock (async) ---

_performSync (polling.js:48)
```

<p>
<img width="390" alt="" src="https://user-images.githubusercontent.com/1623628/79930338-d07f9c80-8422-11ea-9ca6-451e84c5f2b5.png">

</details>

I believe the behaviour described here is as old as 507397f6c (#5431) which first shipped in v4.14.0.

### On `develop` @ [`d90810263`](https://github.com/MetaMask/metamask-extension/tree/d90810263)

Result:

- The transaction is stuck in pending (as expected)
- The `txMeta.history` for the transaction will grow unbounded at one rate:
    - **Rapidly** always
        - An example `txMeta.history`: [`develop-fast-history.log`](https://github.com/MetaMask/metamask-extension/files/4513357/develop-fast-history.log)

This is all worse on `develop` where we essentially hit the pathological case with each pending transaction. As of #8351, `_checkPendingTx` will _always_ emit a warning for a pending transaction. Admittedly this is also related to #7483, where the fallback for a `null` transaction receipt was originally mismanaged.

---

Notes:

- This investigation was spurred on by #6547
- [`v778-slow-history.log`](https://github.com/MetaMask/metamask-extension/files/4513268/v778-slow-history.log)
- [`v778-fast-history.log`](https://github.com/MetaMask/metamask-extension/files/4513339/v778-fast-history.log)
- [`v778-network-fail-stacktrace-pathological.txt`](https://github.com/MetaMask/metamask-extension/files/4513292/network-fail-778-stacktrace-pathological.txt)
- [`develop-fast-history.log`](https://github.com/MetaMask/metamask-extension/files/4513357/develop-fast-history.log)

**Comment by whymarrh** [comment]
cc @danfinlay @rekmarks @Gudahtt

**Comment by rekmarks** [comment]
This is wonderful work!

>The rapid case—the worse one—is primarily the following loop:

Just to confirm: your use of "primarily" here is to indicate that the most relevant steps of the loop are those you describe, not that you have reason to believe that there's some other, unknown cause?

**Comment by whymarrh** [comment]
Yes, I cut a few frames out of the stack trace in describing it

**Comment by whymarrh** [comment]
These fixes are slated for 7.7.9: #8444
